document projection clause changes (#1618)
neo4j/docs-cypher: modules/ROOT/pages/functions/aggregating.adoc

|===

Grouping keys themselves can be complex expressions.
Cypher only recognizes a sub-expression in aggregating expressions as a grouping key if the grouping key is either:

- A variable - e.g. the `p` in `RETURN p, p.age - max(f.age)`.
- A property access - e.g. the `p.age` in `RETURN p.age, p.age - max(f.age)`.
- A map access - e.g. the `p.age` in `WITH {name:'Keanu Reeves', age:58} AS p RETURN p.age, p.age - max(p.age)`.

These are called _recognizable grouping key expressions_.

> If more complex grouping keys are needed as operands in aggregating expression, it is always possible to project them with `AS` in `RETURN` and `WITH`.

Author: ```suggestion
If you need more complex grouping keys as operands in aggregating expressions, you can always project them with `AS` in `RETURN` and `WITH`.
```

* a grouping key, e.g. the `a` in `RETURN a, count(*)`.
* a local variable, e.g. the `x` in  `count(*) + size([ x IN range(1, 10) | x ])`.
* a sub-expression, all operands of which have to be allowed in an aggregating expression.

[role=label--new-2026.07]
[[expression-visibility]]
=== Visibility of expressions in WHERE and ORDER BY

The visibility of expressions in `WHERE` and `ORDER BY` has been streamlined:

> . Projected aliases (variables named with `AS` in `RETURN` and `WITH` clauses) shadow variables defined prior to `RETURN` or `WITH`, if they share the same name, in the scope of subclauses (like `WHERE` and `ORDER BY`).

Reviewer: This is only true for clauses using GROUP BY, without it the behavior is still inconsistent and deprecated. In the future this will be the same for all clauses.

Author: so we should be more clear about the behavior changing with GROUP BY only for now?
when in the future?
